core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java

```
@@ -0,0 +1,279 @@
+
+    public static ShapeBuilder parse(XContentParser parser, GeoShapeFieldMapper shapeFieldMapper)
+            throws IOException, ElasticsearchParseException {
+        FastStringReader reader = new FastStringReader(parser.text());
+        try {
+            // setup the tokenizer; configured to read words w/o numbers
+            StreamTokenizer tokenizer = new StreamTokenizer(reader);
+            tokenizer.resetSyntax();
+            tokenizer.wordChars('a', 'z');
+            tokenizer.wordChars('A', 'Z');
+            tokenizer.wordChars(128 + 32, 255);
+            tokenizer.wordChars('0', '9');
+            tokenizer.wordChars('-', '-');
+            tokenizer.wordChars('+', '+');
+            tokenizer.wordChars('.', '.');
+            tokenizer.whitespaceChars(0, ' ');
+            tokenizer.commentChar('#');
+            return parseGeometry(tokenizer);
+        } finally {
+            reader.close();
+        }
+    }
+
+    /** parse geometry from the stream tokenizer */
+    private static ShapeBuilder parseGeometry(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
+        final GeoShapeType type = GeoShapeType.forName(nextWord(stream));
+        switch (type) {
+            case POINT:
+                return parsePoint(stream);
+            case MULTIPOINT:
+                return parseMultiPoint(stream);
+            case LINESTRING:
+                return parseLine(stream);
+            case MULTILINESTRING:
+                return parseMultiLine(stream);
+            case POLYGON:
+                return parsePolygon(stream);
+            case MULTIPOLYGON:
+                return parseMultiPolygon(stream);
+            case ENVELOPE:
+                return parseBBox(stream);
+        }
+        throw new IllegalArgumentException("Unknown geometry type: " + type);
+    }
+
+    private static EnvelopeBuilder parseBBox(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
```

This will also accept something like "BBOX (-10) 10) -10) 10)", I think this should throw an error. Its a bit constructed but I think the parser shouldn't be that lenient.

---

:+1: